### ISSUE TITLE
Incorrect validation of rule keys substring of schema keys

### ISSUE THREAD

**Issue by MatElGran** [dry-rb/dry-validation#705]
## Describe the bug

Contract#rule does not reject invalid keys if they are substring of valid keys

## To Reproduce

<details>
<summary>
reproduction script
</summary>

```ruby
require 'bundler/inline'
gemfile(true) do
  source 'https://rubygems.org'

  gem 'dry-validation'
end

class Test < Dry::Validation::Contract
  schema do
    required(:tags).array(:hash) do
      required(:name).filled(:string)
    end
  end
  
  rule(:tag) do # This should raise an InvalidKeysError
    key.failure('must have at least 1 element') if value.empty?
  end
end
```

</details>

## Expected behavior

Contract#rule should raise an InvalidKeysError

## My environment

- Affects my production application: NO
- Ruby version: 2.7
- OS: macOS
